### ISSUE TITLE
Silently ignoring changes to default_options in project() is a bad user experience

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#2193]
This has come up repeatedly, most recently when @pwithnall saw CI failures that didn't get fixed when he added `c_std=c99` to `default_options` because we silently ignore changes to it on reconfigure.

Our options are:

1. Change the outputted `build.ninja` to contain changes to `default_options` (tricky; f.ex what if `default_options` changes a value that the user had previously set on the command-line?)
2. Warn that default options were changed, and users should run `ninja wipe-reconfigure` (which is a target that we've talked about which would wipe all meson configuration and re-configure from scratch)

The current situation is somewhat bad, IMO.

**Comment by nirbheek** [comment]
> That’s not useful for the CI use case, where warnings mean nothing to the CI machine

We can add CI-specific behaviour, which is useful in other cases too. See: https://github.com/mesonbuild/meson/pull/1662

> What’s the use case for not reconfiguring when default_options is changed?

I can't think of one; the biggest obstacle is the code itself. We currently don't track whether an option is currently at its 'meson-default' value, its 'build-file-default' value, or has been overridden on the command-line, so if the 'build-file-default' changes, we can't know whether to change the value and regenerate `build.ninja` with it.

The best solution would probably be to fix this.

**Comment by jpakkane** [comment]
Because that is what it is, a _default value_, not _current value_. A design principle in Meson has been that values of options can _never_ be changed from within the build files. The reason for this is that CMake does allow for it and it is maddeningly confusing.

In Meson the process of building must be functional. That is, given the same inputs (source files and option values) it must _always_ produce the same output. Allowing existing options to change values breaks this, and at worst leads to a non-converging build setup where running the build again will always produce different outputs on two consecutive runs.

**Comment by nirbheek** [comment]
> Because that is what it is, a default value, not current value. 

But the default values affect the current value, by definition, if you don't override that option from the command line.

> A design principle in Meson has been that values of options can never be changed from within the build files

I think the `project()` invocation is an exception because we parse it before we start parsing the build files themselves.

> That is, given the same inputs (source files and option values) it must always produce the same output. Allowing existing options to change values breaks this, and at worst leads to a non-converging build setup where running the build again will always produce different outputs on two consecutive runs.

I don't understand this. Changing the build files leads to a reconfigure, and produces the wrong output compared to if you configure with a fresh build directory. Doesn't that lead to a non-converging build setup?

**Comment by jpakkane** [comment]
> But the default values affect the current value, by definition, if you don't override that option from the command line.

They  do not define the current value. What they mean is "in case this was not already defined, then use this value".

>  Changing the build files leads to a reconfigure, and produces the wrong output compared to if you configure with a fresh build directory.

It's not "wrong" as such, just different. The two builds have a different value for an option and thus the output is different (usually).

> Doesn't that lead to a non-converging build setup?

It leads to a different setup, but both of them are stable. What I meant here is if you can change option values (with a hypothetical `set_option`) then someone can do this:

    if get_option('some_bool')
      # do something
    else
      #do something else
    endif
    set_option(not get_option('some_bool'))

This will never converge. Whenever you run the build it will change things and trigger a full reconfigure the next time you run ninja. Or if it doesn't then you have a mismatch between what the value of `some_bool` actually is and what it should be. (This is slightly tangent to the issue here but the same point still stands.)

Default values were added to address exactly this. They provide a way to set up a custom default value but which are ignored on all consecutive runs. This is also in the documentation but sadly people don't read through all of it and if they do then they don't remember all the niggles.

I admit that the way the current thing behaves is a bit confusing the first time you hit it but the alternatives are worse and no-one has yet come up with a way to make this clear and understandable in all the corner cases.

**Comment by nirbheek** [comment]
> They do not define the current value. What they mean is "in case this was not already defined, then use this value".

The value was indeed not set, so why wouldn't it use the new default value now?

> It's not "wrong" as such, just different. The two builds have a different value for an option and thus the output is different (usually).

If the same build files with the same command-line and the same environment lead to two different outputs, that's a bug. It violates user's expectations and hence the principle of least surprise.

> It leads to a different setup, but both of them are stable. What I meant here is if you can change option values (with a hypothetical set_option) then someone can do this:

That's fine, but we don't have `set_option`, and we never will, and this is about `project()` which cannot have conditional default options.

**Comment by jpakkane** [comment]
> The value was indeed not set, so why wouldn't it use the new default value now?

Here "set" means "had some value before Meson started executing any code".

> If the same build files with the same command-line and the same environment lead to two different outputs, that's a bug.

Yes. But that is not the case here. A fresh run has a different environment than a rerun one: it does not have values for options. They don't exist at all in the former but do have pre-existing values in the latter.

> this is about project() which cannot have conditional default options.

Well yes and no. There are all sorts of nasty things one could do such as:

    if get_option()
      subproject('foo', default_options : 'baz=bob')
    else
      subproject('foo', default_options : 'baz=bar')
    endif

Or possibly even:

    project('foo', default_options : get_option('foo') ? 'foo=false' : 'foo=true')

Or if you have changed the value of `foo` from its default value to `bar` and then you have consecutive commits that change the default value first to `bar` and then to `baz`. If you pull both changes in one step the value remains `bar` but if you pull them one at a time, the final result will have value `baz`.

**Comment by jpakkane** [comment]
Storing the changes and warning (#2 in the post) is workable, we can do that.

**Comment by pwithnall** [comment]
> Storing the changes and warning (#2 in the post) is workable, we can do that.

Warning is no use for CI unless there’s an option to turn it into a failure. Even better, rather than failing and requiring human intervention to fix (sad times), it would automatically reconfigure.

I don’t really care about the semantics of how options are handled. I pushed a change to git, and it wasn’t reflected in the next CI run. This should be possible to fix without the fix requiring human intervention.

**Comment by gauteh** [comment]
I agree, options set through `project()` should trigger a reconfig - if they've been explicitly set by the user, e.g. a different test-suite run is one of the few times that will happen, then print a warning - otherwise update the option. If you have to use a `value|default-value` setup that is cumbersome, but as far as I can understand from the comments here `project()` is one of the few places where options can be set both by the build file and the user - perhaps then the project options should be limited to constants and not any evaluations?

**Comment by NickeZ** [comment]
Couldn't we always print the default values during reconfigure? That way you would notice that "these are not the default values i expected". If we implement option 2 in OP, you have to see it in the output of the first reconfiguration. Always printing the default values would let you see the inconstancy every time.

**Comment by jpakkane** [comment]
We already print a lot of stuff, I fear this would just add noise that people would almost always ignore.

**Comment by NickeZ** [comment]
What about printing all options that are not the default values every time?

**Comment by nioncode** [comment]
I agree that for a CI environment (or even most default developer machines) where options are never set manually with `meson configure` and just the defaults are used, changing the `default_options` in `meson.build` has to trigger a reconfigure, just like it does when changing the option with `meson configure`. Otherwise, somebody has to clean the build directory on the CI or manually update the options in the build directory.

For this to work, it should be enough to store a flag for each option that denotes if it has been changed manually by the user through `meson configure` or if it is the meson / project default. Then, when the options are changed in `meson.build`, this flag can be checked to see if the value should be updated from `default_options`.

**Comment by jpakkane** [comment]
> Otherwise, somebody has to clean the build directory on the CI or manually update the options in the build directory.

CI should always run from a clean slate. That's what it's _for_ and also why CCache was invented.

> Then, when the options are changed in meson.build, this flag can be checked to see if the value should be updated from default_options.

Which means that whether something changes value or not is based on hidden state. Which is also an awful user experience.

**Comment by nioncode** [comment]
meson's approach has always been that you configure your build dir *once* and only have to call ninja afterwards. However, this is not true if you modify the `default_options` of your project. Then you either have to wipe your build dir or also change the option through `meson configure`.

*This* is bad user experience. If I never touched the default values, but then change them in my `meson.build`, I'd expect meson to pick this up and reconfigure accordingly.

**Comment by xclaesse** [comment]
> For this to work, it should be enough to store a flag for each option that denotes if it has been changed manually by the user through meson configure or if it is the meson / project default.

FWIW, that's already stored in the cmd_line.txt file, only options there have been modified by the user are stored there. If you reconfigure with `--wipe` it will update to new `default_options` unless you have set a value to that option. But it has the downside that you'll have to rebuild everything from scratch. I'm wondering if we should have something like a `--soft-wipe` that would nuke only `coredata.dat` to make a reconfigure from scratch (loading new default_options) but won't delete all built objects to avoid a full rebuild.

**Comment by nioncode** [comment]
The thing is, as a user I don't want to call any other command manually after updating my build files. I simply want to call `ninja` and everything should work and not use old information that I never touched myself.

**Comment by egnor** [comment]
This is still befuddling noobs (me) several years after the initial report. Is there a best practice for avoiding this extremely confusing problem?

**Comment by nyanpasu64** [comment]
I tried out these commands and they seem to work properly (ninja doesn't try to rebuild all files, not even run ccache):

```sh
cd builddir
rm meson-private/coredata.dat
meson setup . ..
```

Is there any reason this should not be added as `meson setup --default-options` or similar? (IMO it's confusing that `--wipe` does *not* wipe user-specified command-line options, which are stored in builddir/meson-private/cmd_line.txt.)

Note that adding that does *not* remove the pitfall for users. Making Meson "just work" by default would require automatically removing coredata.dat whenever default_options is changed. I'm not sure how to detect it changing, perhaps either by caching default_options and comparing the new meson.build contents, or by always regenerating coredata.dat whenever meson.build changes and applying the new settings if they differ.

**Comment by ishitatsuyuki** [comment]
I just stumbled upon this lately, and as other say I don't think this behavior makes sense. I expect `meson setup --reconfigure` to be "generate build.ninja as if we were starting fresh (but do reuse build artifacts if they're built with the same arguments), except remember all the command line flags that have been passed in before".

I don't get the argument about "convergence" in https://github.com/mesonbuild/meson/issues/2193#issuecomment-322566536 either. The way I just described will converge no worse, since it's mostly stateless except the remembered command line flags.

Is there any reason we can't move to something like I (and some others) described? The implementation would not be that hard either, as a starting point we can just discard `coredata.dat`, which is a small hit in reconfigure performance but gives much less hassle.

**Comment by matty0ung** [comment]
It would be good to understand why so many of these settings are available from the command line.  I get that you might want to make a command line choice about which directory to build in or whether to do a debug or a release build.  But it's surprising to me that, eg, choosing between C++14, C++17 and C++20 isn't exclusively set in `meson.build`. Could someone give an example of when that's needed?

**Comment by iameli-streams** [comment]
Fighting with getting GStreamer compiled into a binary and this has lost me a lot of time. Contemplating writing something that parses meson.build, pulls out all the default options, and applies them via `meson configure`.

Right now my workflow is like... I need to change some settings for the GStreamer subproject. I need to switch its `gst-full-elements` to `coreelements:fdsrc,fdsink,queue`. My options are...

1. Change the `default_options` for the subproject in meson.build, delete `build`, and trigger a full rebuild. Takes a while. CCache doesn't help all that much in my project, due to Go and Rust dependencies.
2. Run `meson configure build '-Dgstreamer-full:gst-full-elements=coreelements:fdsrc,fdsink,queue'`. Works immediately, but then I'll forget to put it into meson.build, so my builds will work great until next time I trigger a full rebuild for whatever reason, and I'll be stuck looking through my fish history trying to figure out what I set when.
3. Do both, which is time-consuming and aggravating. I've been trying to do this, but the syntax for the two is completely different. So I copy paste `'-Dgstreamer-full:gst-full-elements=coreelements:fdsrc,fdsink,queue'` into my meson.build, delete the `=` in the middle, replace it with a `:`, delete the `-Dgstreamer-full:`, surround both sides in single quotes... and then my build fails anyway because I forgot to include the trailing comma.

Repeat dozens of times as I tweak the dependencies to try and get things building properly. Not fun. It's got me in a place where I'm seriously considering just putting a lot of `-D` clauses in a Makefile and handle all of the options entirely outside of `meson.build`. That'd work really elegantly, actually! Reconfigures would be really fast and fresh builds would have all the options. But it's probably a bad sign that I find myself needing to work outside of Meson to get this working.

Something like:
```
meson configure --reapply-defaults
```
would fix my problem instantly.

**Comment by iameli-streams** [comment]
I did it. Truly, this is the future of compilation!

```Makefile
.PHONY: meson-setup
meson-setup:
	meson setup build --reconfigure \
		-D "gst-plugins-base:audioresample=enabled" \
		-D "gst-plugins-base:playback=enabled" \
		-D "gst-plugins-base:opus=enabled" \
		-D "gst-plugins-base:gio-typefinder=enabled" \
		-D "gst-plugins-base:typefind=enabled" \
		-D "gst-plugins-good:matroska=enabled" \
		-D "gst-plugins-good:soup=enabled" \
		-D "gst-plugins-bad:fdkaac=enabled" \
		-D "gstreamer-full:gst-full=enabled" \
		-D "gstreamer-full:gst-full-plugins=libgstaudioresample.a;libgstmatroska.a;libgstfdkaac.a;libgstopus.a;libgstplayback.a;libgstsoup.a" \
		-D "gstreamer-full:gst-full-libraries=gstreamer-controller-1.0,gstreamer-plugins-base-1.0,gstreamer-pbutils-1.0" \
		-D "gstreamer-full:gst-full-target-type=static_library" \
		-D "gstreamer-full:gst-full-elements=coreelements:fdsrc,fdsink,queue,typefind" \
		-D "gstreamer-full:bad=enabled" \
		-D "gstreamer-full:ugly=disabled" \
		-D "gstreamer-full:tls=disabled" \
		-D "gstreamer-full:gpl=enabled"
```

I poke fun because I love Meson so much; truly some of what I'm attempting here would be infeasible without it. But I'm positive we can come up with a better UX here.

**Comment by eli-schwartz** [comment]
> Do both, which is time-consuming and aggravating. I've been trying to do this, but the syntax for the two is completely different. So I copy paste `'-Dgstreamer-full:gst-full-elements=coreelements:fdsrc,fdsink,queue'` into my meson.build, delete the `=` in the middle, replace it with a `:`, delete the `-Dgstreamer-full:`, surround both sides in single quotes... and then my build fails anyway because I forgot to include the trailing comma.

The syntax should be exactly the same. `-Dvar=val` maps to `default_options: ['var=val']`.

So:
- copy-paste `'-Dgstreamer-full:gst-full-elements=coreelements:fdsrc,fdsink,queue'` into your meson.build
- delete the "-Dgstreamer-full"

Meson doesn't require trailing commas unless they disambiguate multiple elements.

**Comment by iameli-streams** [comment]
Sure, that's somewhat better - I was using a dictionary previously, so this was what I was replacing:
```
gstreamer_proj = subproject(
   'gstreamer-full',
   default_options: {
     'gst-full': 'enabled',
     'gst-full-plugins': 'libgstaudioresample.a;libgstmatroska.a;libgstfdkaac.a;libgstopus.a',
     'gst-full-libraries': 'gstreamer-controller-1.0',
     'gst-full-target-type': 'static_library',
     'bad': 'enabled',
     'ugly': 'disabled',
     'tls': 'disabled',
     'gpl': 'enabled',
     'force_fallback_for': 'glib,gio',
     'gst-full-elements': 'coreelements:fdsrc,fdsink',
   },
)
```
But I can see how a list would be a bit easier for me here.

**Comment by eli-schwartz** [comment]
The general topic here has been discussed in multiple issues over the years. As has been said elsewhere (not sure offhand where) it is probably a good idea for meson to start tracking whether individual values have been set from the command line, so that changes to the default value can be applied automatically when the user hasn't explicitly overridden it.


Any solution that causes `meson setup builddir -Dfoo=1 && git pull && ninja -C builddir` to reconfigure the project and reset the value of option "foo" to "0" is not likely to happen. If you specified a value on the command line then things should not change.

**Comment by eli-schwartz** [comment]
> Sure, that's somewhat better - I was using a dictionary previously, so this was what I was replacing:

Ah yeah, I see what you mean now.

**Comment by bonzini** [comment]
This actually has a chance of being fixable with the option refactoring. "Just" move all nondefault setting of the options to `optstore.augments` instead of overwriting the option object.

**Comment by Hi-Angel** [comment]
So, while this is not fixed, what is user supposed to do, barring deleting build dir completely? I kind of assumed a `meson setup build --reconfigure` should, well, reconfigure the build — however after reconfiguration meson still uses the options no longer existing.

**Comment by nyanpasu64** [comment]
```sh
cd builddir
rm meson-private/coredata.dat*
ninja
```

You could try playing around with similar commands, this is what worked for me in my last test.
